src/findex.rs

```diff
@@ -0,0 +1,262 @@
+use std::{
+    collections::{HashMap, HashSet},
+    hash::Hash,
+    sync::{Arc, Mutex},
+};
+
+use tiny_keccak::{Hasher, Sha3};
+
+use crate::{
+    ADDRESS_LENGTH, Address, IndexADT, KEY_LENGTH, MemoryADT, adt::VectorADT, encoding::Op,
+    encryption_layer::MemoryEncryptionLayer, error::Error, ovec::IVec, secret::Secret,
+};
+
+#[derive(Clone, Debug)]
+pub struct Findex<
+    const WORD_LENGTH: usize,
+    Value,
+    TryFromError: std::error::Error,
+    Memory: Send + Sync + Clone + MemoryADT<Address = Address<ADDRESS_LENGTH>, Word = [u8; WORD_LENGTH]>,
+> where
+    // values are serializable (but do not depend on `serde`)
+    for<'z> Value: TryFrom<&'z [u8], Error = TryFromError> + AsRef<[u8]>,
+    Memory::Error: Send + Sync,
+{
+    el: MemoryEncryptionLayer<WORD_LENGTH, Memory>,
+    cache: Arc<
+        Mutex<
+            HashMap<
+                Address<ADDRESS_LENGTH>,
+                IVec<WORD_LENGTH, MemoryEncryptionLayer<WORD_LENGTH, Memory>>,
+            >,
+        >,
+    >,
+    encode: Arc<
+        fn(
+            Op,
+            HashSet<Value>,
+        )
+            -> Result<Vec<<MemoryEncryptionLayer<WORD_LENGTH, Memory> as MemoryADT>::Word>, String>,
+    >,
+    decode: Arc<
+        fn(
+            Vec<<MemoryEncryptionLayer<WORD_LENGTH, Memory> as MemoryADT>::Word>,
+        ) -> Result<HashSet<Value>, TryFromError>,
+    >,
+}
```

Comment:
Impressive!
Any way to simplify it? (by creating alias type for example)
```
type EncryptedIVec<const WORD_LENGTH: usize, Memory> =
    IVec<WORD_LENGTH, MemoryEncryptionLayer<WORD_LENGTH, Memory>>;
type EncryptedVec<const WORD_LENGTH: usize, Memory> =
    Vec<<MemoryEncryptionLayer<WORD_LENGTH, Memory> as MemoryADT>::Word>;
```
or any type that works for you. But it's pretty hardcore to read.
Also `Op` could be renamed in `Operation` I suppose